### PR TITLE
Create a dh key to fix ssl bind of ejabberd

### PR DESCRIPTION
**Pull request by stephdl**
When you bind a remote ejabberd server to your openldap server over ssl (636 tcp), ejabberd gets a TLS handshake error, the usage of a dh key fix the the issue.

following https://access.redhat.com/articles/1474813 RHEL proposes to remove the usage of `TLS_DHE_RSA` which are enabled by the diffie hellman key see `!EDH` however the `TLS_DHE_RSA`  ciphers are mandatory to enable the ssl bind of ejabberd

see the diff of cipher, we add more ciphers `TLS_DHE_RSA*` this are mandatory for ejabberd

https://gist.github.com/stephdl/984fdf81a9dd177bb7c7723833930d22

https://github.com/NethServer/dev/issues/6409

### CODE REVIEW

File: root/etc/e-smith/events/actions/nethserver-directory-tlspolicy [hunk 1]
```diff
@@ -20,14 +20,18 @@
...
+#
+
 tlspolicy=$(/sbin/e-smith/config getprop tls policy)
 
 if (( ${tlspolicy:-0} >= 20200510 )); then
     ldapmodify -c -Y EXTERNAL &>/dev/null <<'EOF' || ((errors ++))
 dn: cn=config
 changetype: modify
 replace: olcTLSCipherSuite
-olcTLSCipherSuite: EECDH:EDH:CAMELLIA:ECDH:RSA:!eNULL:!SSLv2:!RC4:!DES:!EXP:!SEED:!IDEA:!3DES
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-directory-tlspolicy:30]: :+1: From https://gist.github.com/stephdl/984fdf81a9dd177bb7c7723833930d22 it seems this change just adds some new ciphers to those already available. In this case we can just fix the latest TLS policy.

Let me check that diff by my side too during the QA ;)

**Review by DavidePrincipi — CHANGES_REQUESTED** [review]
Generating the DH parameter file is time consuming for systems with poor entropy generation. 

I prefer to follow the approach of https://github.com/NethServer/nethserver-mail/pull/203

Shall we provide a default DH params file to keep Slapd running while we wait for the new file?

**Comment by stephdl** [comment]
> Shall we provide a default DH params file to keep Slapd running while we wait for the new file?

Unfortunately openldap doesn't have a default DH key, so no :( However the only interest in your idea is if we remove the key (a DH key is not here to stay) you just have to restart the service to get a new one.

So yes we need either an event or a service start to create the DH key. My only concern is that at least one time you have to wait between 10 seconds to 60 seconds for the creation. I bet for openvpn it is the same, the key needs to be created at least one time

**Comment by Neustradamus** [comment]
@stephdl: Thanks for your improvement!

Like I have said some months ago, do not forget to add a DH key into ejabberd too:
- https://github.com/NethServer/nethserver-mail/pull/203#issuecomment-650508233

**Comment by stephdl** [comment]
@Neustradamus not the goal of the PR but I will think on it

@DavidePrincipi we have a tiny issue

The asynchronous service is started by the action of the update event no issue on this side

**IF on a server that the ldapmodify has been done already for `olcTLSDHParamFile`**
- you remove the dh key 
- start the event directory-update
- the service slapd fail to start because the dh key is missing (my tiny issue)
- we launch the service slapd-create-DH2048.service
- if the slapd is in failed state we start it (a trick I added)
- we launch the ldapmodify again (no issue on this side)
- at the end slapd is up

so in short to change  the DH key

rm the key
signal-event nethserver-directory-update

you will have probably some fails in /var/log/message but the new service once the key will be created will start again slapd

File: root/usr/lib/systemd/system/slapd-create-DH2048.service [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[Unit]
+Description=Create the DH2048 group for slapd
+ConditionPathExists=!/etc/openldap/certs/slapd.dh.params
+
+[Service]
+Type=simple
+ExecStart=/usr/libexec/nethserver/slapd-create-DH2048
```
Comment (DavidePrincipi) [inline root/usr/lib/systemd/system/slapd-create-DH2048.service:7]: The number of bits may be changed in the future: I'd prefer a "neutral" name

```suggestion
ExecStart=/usr/libexec/nethserver/slapd-create-dhparam
```

File: root/usr/libexec/nethserver/slapd-create-DH2048 [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+tmpfile=$(mktemp /tmp/slapd-create-DH2048.XXXXXXX)
+
+/usr/bin/openssl dhparam -out ${tmpfile} 2048 2> /dev/null || ((errors ++))
+/usr/bin/mv -f ${tmpfile} /etc/openldap/certs/slapd.dh.params || ((errors ++))
+/usr/bin/chown root:ldap /etc/openldap/certs/slapd.dh.params &>/dev/null || ((errors ++))
+/usr/bin/chmod 440 /etc/openldap/certs/slapd.dh.params &>/dev/null || ((errors ++))
+
+# Ensure that slapd is running if the DH key has been deleted and the service failed to start
+systemctl -q is-failed  slapd && systemctl start slapd
```
Comment (DavidePrincipi) [inline root/usr/libexec/nethserver/slapd-create-DH2048:33]: If slapd is not running this script must fail

```suggestion
```

**Review by DavidePrincipi — COMMENTED** [review]
The unit must fail if its precondition (slapd is running) is not satisfied. We need to analyze also what happens during the "first installation" of nethserver-directory and the "restore from backup" use cases. There could a race condition with other LDAP initialization and provisioning actions. 

We have to evaluate to include the DH file in the config-backup. As alternative remove `olcTLSDHParamFile` from the .ldif dump file of the LDAP config DB.

----

Let's establish consistent naming rules for the action, unit, helper script, and output file. I suggest:

1. action: `nethserver-directory-create-dhparam`
2. unit: `slapd-create-dhparam.service`
3. helper script: `slapd-create-dhparam`

File: root/usr/lib/systemd/system/slapd-create-DH2048.service (hunk 2 above)
Comment (Neustradamus) [reply root/usr/lib/systemd/system/slapd-create-DH2048.service:7]: @DavidePrincipi: I think that @stephdl has based the name with:
- https://github.com/NethServer/nethserver-mail/blob/master/common/usr/lib/systemd/system/postfix-create-DH2048.service
- https://github.com/NethServer/nethserver-mail/blob/master/common/usr/libexec/nethserver/postfix-create-DH2048

Comment (Neustradamus) [reply root/usr/lib/systemd/system/slapd-create-DH2048.service:7]: Note: Currently:
- https://github.com/search?q=org%3ANethServer+DH1024&type=code: nethserver-mail + nethserver-vpn + nethserver-openvpn
- https://github.com/search?q=org%3ANethServer+DH2048&type=code: nethserver-mail + nethserver-openvpn

**Comment by stephdl** [comment]
Well or go back to my first idea, create the dh key in nethserver-directory-update event eventually make a condition before to start slapd that the key must be created, even if we do need to wait a time, we are sure we never fail to start slapd.

when you install the update, when you do a restoration, the event is launched and the key is created

File: root/etc/e-smith/events/actions/nethserver-directory-tlspolicy [hunk 4]
```diff
@@ -20,14 +20,18 @@
 # along with NethServer.  If not, see COPYING.
 #
 
+#
+# Following REHL recommendation https://access.redhat.com/articles/1474813
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-directory-tlspolicy:24]: Typo

```suggestion
# Following RHEL recommendation https://access.redhat.com/articles/1474813
```

File: root/etc/e-smith/events/actions/nethserver-directory-conf [hunk 5]
```diff
@@ -33,6 +33,13 @@ if ( ! -f '/var/lib/ldap/cn.bdb' ) { # first configuration
     }
 }
 
+#
+# FIXME : slapd is not restarted by the event update, 
+# then systemd configuration for the service 
+# slapd-create-dhparam.service is not reloaded
+#
+system('/usr/bin/systemctl daemon-reload');
+
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-directory-conf:42]: The PR adds a .conf fragment and a .service definition: use RPM macros for systemd, according to Fedora RPM packaging guidelines.

```suggestion
```

File: nethserver-directory.spec [hunk 6]
```diff
@@ -44,6 +44,8 @@ rm -rf %{buildroot}
 %ghost %attr(644,root,root) /etc/pam.d/password-auth-nh
 %ghost %attr(644,root,root) /etc/pam.d/system-auth-nh
 %config %attr (0440,root,root) %{_sysconfdir}/sudoers.d/20_nethserver_directory
+%attr (0750,root,root) /usr/libexec/nethserver/slapd-create-dhparam
```
Comment (DavidePrincipi) [inline nethserver-directory.spec:47]: Our RPM macro should set 0755 perms and the binary is outside of PATH settings. I think there's no practical need to restrict the executable permissions for normal users.

```suggestion
```

File: nethserver-directory.spec [hunk 7]
```diff
@@ -44,6 +44,8 @@ rm -rf %{buildroot}
...
+%config(noreplace) %attr (0440,root,ldap) /etc/openldap/certs/slapd.dh.params
```
Comment (DavidePrincipi) [inline nethserver-directory.spec:48]: I'd really like a standard file extension. As the file is in PEM format, the extension can be .pem: what else?

```suggestion
%config(noreplace) %attr (0440,root,ldap) /etc/openldap/certs/slapd-dhparam.pem
```

File: root/usr/libexec/nethserver/slapd-create-dhparam [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+if /usr/bin/grep -q '3TGqya1beM2QPhXUPLYCdNrz+dl6qftStqWNc4z7BRl8PeGzun/oDsTosPuXDhCf' /etc/openldap/certs/slapd.dh.params 2>/dev/null;then
+    tmpfile=$(mktemp /tmp/slapd-create-dhparam.XXXXXXX)
+
+    /usr/bin/openssl dhparam -out ${tmpfile} 2048 2> /dev/null || ((errors ++))
+    /usr/bin/mv -f ${tmpfile} /etc/openldap/certs/slapd.dh.params || ((errors ++))
+    /usr/bin/chown root:ldap /etc/openldap/certs/slapd.dh.params &>/dev/null || ((errors ++))
+    /usr/bin/chmod 440 /etc/openldap/certs/slapd.dh.params &>/dev/null || ((errors ++))
+
+exit $errors
+fi
```
Comment (DavidePrincipi) [inline root/usr/libexec/nethserver/slapd-create-dhparam:32]: 1. exit early if no action is needed,  exit early on error: `set -e`
1. do not capture stdout/stderr: the journal may be interesting
1. the long DH generation could be stopped by external actions (system reboot?) and /tmp is persistent across reboots. Ensure tmpfile is always removed
1. why not adding a custom section (NETHSERVER DEFAULT) for matching in the default PEM file? The random string is scary
1. change the file with .pem suffix :)

```suggestion
if ! /usr/bin/grep -q 'NETHSERVER DEFAULT' /etc/openldap/certs/slapd.dh.params; then
    exit 0
fi

tmpfile=$(mktemp /tmp/slapd-create-dhparam.XXXXXXX)
trap "/usr/bin/rm -f ${tmpfile}" EXIT

set -e # Exit immediately if any error occurs

/usr/bin/openssl dhparam -out ${tmpfile} 2048
/usr/bin/mv -f ${tmpfile} /etc/openldap/certs/slapd.dh.params
/usr/bin/chown root:ldap /etc/openldap/certs/slapd.dh.params
/usr/bin/chmod 440 /etc/openldap/certs/slapd.dh.params
```

Comment (stephdl) [reply root/usr/libexec/nethserver/slapd-create-dhparam:32]: >  why not adding a custom section (NETHSERVER DEFAULT) for matching in the default PEM file? The random string is scary

I tested it ... and slapd did not like it
the ldapmodify we do in the action nethserver-directory-conf was broken, I spent some time just to understand that when you modify  something inside the key file, you change the checksum. We need to grep a part of the key, it is the same

**Comment by stephdl** [comment]
Due to the systemd issue, the daemon of systemd is not reloaded, the service to create the key is not started during the rpm installation, only a `systemctl restart slapd` after the `runlevel-adjust`

possible way

- let this as is, the next restart of slapd will create the key
- add in %post a daemon-reload or a start for `slapd-create-dhparam.service`
- create an esmith action to launch the service start for `slapd-create-dhparam.service`

**Review by DavidePrincipi — APPROVED** [review]
Thank you for your patience and the hard work!

**Comment by DavidePrincipi** [comment]
Final notes

- systemd daemon-reload was not added. Updated systems will use the default dhparam file until next reboot, or other package triggers the daemon-reload and slapd is restarted. This is not an issue as dhparams are exchanged during TLS handshake
- to avoid subtle issues with the openssl implementation in slapd the dhparams file name is exactly the one suggested by the RHEL kb article. The .pem extension was reverted
